### ISSUE TITLE
MISSING_RESPONSE_SCHEMA warnings

### ISSUE THREAD

**Issue by carlskii** [IBM/openapi-to-graphql#390]
Hi, 

Im trying to use `openapi-to-graphql` to convert this OAS to graphql -https://app.swaggerhub.com/apis/OpenChannel/Market-API/2.0.25

I've download the schema and I'm using this command to process it: 

`openapi-to-graphql OpenChannel-Market-API-2.0.25-resolved.yaml`

However, I'm getting multiple warnings as follows: 

```
"warnings": [
    {
      "type": "MISSING_RESPONSE_SCHEMA",
      "message": "Operation GET /apps has no (valid) response schema. You can use the fillEmptyResponses option to create a placeholder schema",
      "mitigation": "Ignore operation."
    },
    {
      "type": "MISSING_RESPONSE_SCHEMA",
      "message": "Operation POST /apps has no (valid) response schema. You can use the fillEmptyResponses option to create a placeholder schema",
      "mitigation": "Ignore operation."
    },
    {
      "type": "MISSING_RESPONSE_SCHEMA",
      "message": "Operation GET /apps/{appId} has no (valid) response schema. You can use the fillEmptyResponses option to create a placeholder schema",
      "mitigation": "Ignore operation."
    } .......

,
  "numOps": 86,
  "numOpsQuery": 34,
  "numOpsMutation": 52,
  "numOpsSubscription": 0,
  "numQueriesCreated": 0,
  "numMutationsCreated": 0,
  "numSubscriptionsCreated": 0
```

I think the problem is related to the sections of the OAS file that look like this:

```
"responses": {
          "400": {
            "description": "Bad Request - Required parameters are missing, malformed or invalid"
          },
          "default": {
            "description": "OK",
            "schema": {
              "$ref": "#/definitions/AppPages"
            }
          }
        }
      },
```
i.e. the `$ref` statement and it not linking to the definitions section. 

Is there anything I can do to resolve this ?

**Comment by carlskii** [comment]
In addition to this I've used `https://github.com/go-swagger/go-swagger` to validate the swagger spec file: 

```
./swagger validate OpenChannel-Market-API-2.0.25-resolved.yaml
2021/04/15 09:53:42
The swagger spec at "OpenChannel-Market-API-2.0.25-resolved.yaml" is valid against swagger specification 2.0
```
I then used the `./swagger expand OpenChannel-Market-API-2.0.25-resolved.yaml` expand command to resolve all $ref instances. However, I still get the same warnings.

**Comment by Alan-Cha** [comment]
@carlskii You are receiving these errors because this API specification is using the `default` HTTP code instead of a standard `200` HTTP code or `2XX`.

According to [OAS](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#responses-object-example), "A 200 response for a successful operation and a default response for others (implying an error)".

Secondly, I believe the schema is supposed to be contained in media encoding object. For more information, look [here](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#response-object).

```json
"responses": {
          "400": {
            "description": "Bad Request - Required parameters are missing, malformed or invalid"
          },
          "2XX": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/definitions/AppPages"
                }
              }
            }
          }
        }
```

After addressing these two issues, OtG seems to output correctly.

Let me know what you think!

**Comment by carlskii** [comment]
Thanks very much for this - its got me much further forwards. However, seem to have a content type issue now: 

```
{
  "data": null,
  "errors": [
    {
      "path": [
        "viewerBasicAuth",
        "app"
      ],
      "message": "Unexpected error value: \"Operation GET /apps/{appId} should have a content-type '*/*' but has 'application/json;charset=UTF-8' instead\"",
      "locations": [
        {
          "line": 1,
          "column": 122
        }
      ]
    }
  ]
}
```

I've tried using the --header option, but it doesn't to seem to make a difference.

**Comment by Alan-Cha** [comment]
@carlskii We currently do not support `*/*` but it will be soon. For now, assuming that this API uses JSON, if you can replace `*/*` with `application/json`, it should work better.

**Comment by carlskii** [comment]
OK thanks. Where would I specify the application/json? In the OAS spec file or somewhere else ?

**Comment by Alan-Cha** [comment]
@carlskii Yes, in the OAS. Wherever `*/*` is used as a media type, if you can replace it with `application/json`, OtG should work better. If you need more help, you can join us on [our Slack channel](graphql.slack.com/messages/openapi-to-graphql) #openapi-to-graphql in the GraphQL workspace (we just created it so there is little history but I will be checking).

**Comment by carlskii** [comment]
Thanks that’s really helpful - thanks.

**Comment by carlskii** [comment]
So I took a look at the OAS spec file. It doesn't seem to specify any content types of */*. Do I need to create a content type somewhere here?

**Comment by Alan-Cha** [comment]
@carlskii I am so sorry for the confusing advice. Let me explain. OtG uses the module [swagger2graphql](https://www.npmjs.com/package/swagger2openapi) to transform Swaggers (OAS 2) into OAS 3s. Part of this translation includes adding a `*/*` media type if none is provided. I have been playing around with OtG, trying to get your Swagger to work and I assumed the `*/*` was a part of your Swagger but it was actually being added by swagger2graphql.

Please take a look at [this](https://gist.github.com/Alan-Cha/cbd34d9a75f1a3da3ad09f551f59e2f7). I used swagger2graphql to translate your Swagger into OAS 3 as well as changed `default` to `2XX` as well as `*/*` to `application/json`. If you use this, OtG is able to produce more along the desired output. See image.

<img width="656" alt="Screen Shot 2021-04-18 at 10 55 44 AM" src="https://user-images.githubusercontent.com/14318081/115150041-a7ddb300-a034-11eb-990e-d0c447405b79.png">

Alternatively, you can simply run 

```bash
openapi-to-graphql https://gist.githubusercontent.com/Alan-Cha/cbd34d9a75f1a3da3ad09f551f59e2f7/raw/927b000a64aa49959777a86d8b329d0fec8dbd6e/openchannel_modified.json
```

which fetches the modified OAS.

Please note that OtG will still produce a number of warnings but do not be alarmed. OtG warnings have varying severity. However, I did notice that some warnings (and solutions) like the `fillEmptyResponses` option is not working as expected. I'm currently trying to fix them. (This is why it's so helpful for users to report issues with their own real-world OASs!)

However, I hope that the modified OAS is enough for you to get started and to try out OtG. I am very sorry for the issues and I promise they will be resolved soon.

**Comment by carlskii** [comment]
Thanks Alan, I will try this. Thanks again for you help with this. It's much appreciated.

**Comment by carlskii** [comment]
Excellent work. This is now working as expected. Thank you very much for helping out with this, its going to be an important part of a project moving forward.

**Comment by carlskii** [comment]
Hi one thing I've noticed is that there doesn't seem to be the equivalent of GET /apps in the mapped GraphQL schema. i.e. the ability to get all apps. It appears it only provide the GET /apps/{appId} whereby it requires an app ID to provided. Is there any way to get this fixed or am I missing something ?
